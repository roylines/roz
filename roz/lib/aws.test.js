const aws = require('./aws');
const AWS = require('aws-sdk');

jest.mock('aws-sdk');
jest.mock('lambda-log');

describe('aws', () => {
  process.env.BASTION_AUTO_SCALING_GROUP = 'ASG';
  describe('autoscaling', () => {
    describe('get', () => {
      it('can call get (no instances)', async () => {
        // conditions
        AWS.AutoScaling()
          .describeAutoScalingGroups()
          .promise.mockResolvedValue({
            AutoScalingGroups: [{DesiredCapacity: 42, Instances: []}],
          });

        // call get
        const response = await aws.autoscaling.get('describeAlarms');

        // assertions
        expect(response).toMatchSnapshot();
        expect(
          AWS.AutoScaling().describeAutoScalingGroups.mock.calls,
        ).toMatchSnapshot();
      });
      it('can call get (instance not ready)', async () => {
        // conditions
        AWS.AutoScaling()
          .describeAutoScalingGroups()
          .promise.mockResolvedValue({
            AutoScalingGroups: [
              {
                DesiredCapacity: 42,
                Instances: [{InstanceId: 'IID'}],
              },
            ],
          });

        AWS.EC2()
          .describeInstances()
          .promise.mockResolvedValue({
            Reservations: [
              {
                Instances: [],
              },
            ],
          });

        // call get
        const response = await aws.autoscaling.get('describeAlarms');

        // assertions
        expect(response).toMatchSnapshot();
        expect(
          AWS.AutoScaling().describeAutoScalingGroups.mock.calls,
        ).toMatchSnapshot();
        expect(AWS.EC2().describeInstances.mock.calls).toMatchSnapshot();
      });
      it('can call get (instance ready)', async () => {
        // conditions
        AWS.AutoScaling()
          .describeAutoScalingGroups()
          .promise.mockResolvedValue({
            AutoScalingGroups: [
              {
                DesiredCapacity: 42,
                Instances: [{InstanceId: 'IID'}],
              },
            ],
          });

        AWS.EC2()
          .describeInstances()
          .promise.mockResolvedValue({
            Reservations: [
              {
                Instances: [
                  {
                    State: {
                      Name: 'STATUS',
                    },
                    LaunchTime: 'LAUNCHTIME',
                    PublicIpAddress: 'IPADDRESS',
                  },
                ],
              },
            ],
          });

        // call get
        const response = await aws.autoscaling.get('describeAlarms');

        // assertions
        expect(response).toMatchSnapshot();
        expect(
          AWS.AutoScaling().describeAutoScalingGroups.mock.calls,
        ).toMatchSnapshot();
        expect(AWS.EC2().describeInstances.mock.calls).toMatchSnapshot();
      });
    });
    describe('set', () => {
      it('should set desired capacity', async () => {
        //call set
        await aws.autoscaling.set({capacity: 42});
        expect(
          AWS.AutoScaling().setDesiredCapacity.mock.calls,
        ).toMatchSnapshot();
      });
    });
  });
  describe('lambda', () => {
    describe('invoke', () => {
      it('should call invoke correctly', async () => {
        //call set
        await aws.lambda.invoke({arn: 'ARN', message: {a: 'A'}});
        expect(AWS.Lambda().invoke.mock.calls).toMatchSnapshot();
      });
    });
  });
  describe('ssm', () => {
    describe('getTelegramSecrets', () => {
      it('should call getParameters correctly', async () => {
        AWS.SSM()
          .getParameters()
          .promise.mockResolvedValue({
            Parameters: [
              {
                Name: 'NAME-telegram-token',
                Value: 'TOKEN',
              },
              {
                Name: 'NAME-telegram-user',
                Value: 'USER',
              },
            ],
          });

        process.env.NAME = 'NAME';
        //call set
        const response = await aws.ssm.getTelegramSecrets();
        expect(response).toMatchSnapshot({token: 'TOKEN', user: 'USER'});
        expect(AWS.SSM().getParameters.mock.calls).toMatchSnapshot();
      });
      it('should cache', async () => {
        //call set
        const response = await aws.ssm.getTelegramSecrets();
        expect(response).toMatchSnapshot({token: 'TOKEN', user: 'USER'});
      });
    });
    describe('getInitialised', () => {
      it('should call getParameter correctly', async () => {
        AWS.SSM()
          .getParameter()
          .promise.mockResolvedValue({
            Parameter: {
              Value: 'false',
            },
          });

        process.env.NAME = 'NAME';
        //call set
        const response = await aws.ssm.getInitialised();
        expect(response).toEqual(false);
        expect(AWS.SSM().getParameter.mock.calls).toMatchSnapshot();
      });
      it('should cache', async () => {
        //call set
        const response = await aws.ssm.getInitialised();
        expect(response).toEqual(false);
      });
    });
    describe('setInitialised', () => {
      it('should call putParameter correctly', async () => {
        process.env.NAME = 'NAME';
        await aws.ssm.setInitialised();
        expect(AWS.SSM().putParameter.mock.calls).toMatchSnapshot();
        const response = await aws.ssm.getInitialised();
        expect(response).toEqual(true);
      });
    });
  });
});
