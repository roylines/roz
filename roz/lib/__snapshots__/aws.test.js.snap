// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aws autoscaling get can call get (instance not ready) 1`] = `
Object {
  "instance": null,
}
`;

exports[`aws autoscaling get can call get (instance not ready) 2`] = `
Array [
  Array [],
  Array [
    Object {
      "AutoScalingGroupNames": Array [
        "ASG",
      ],
    },
  ],
  Array [],
  Array [
    Object {
      "AutoScalingGroupNames": Array [
        "ASG",
      ],
    },
  ],
]
`;

exports[`aws autoscaling get can call get (instance not ready) 3`] = `
Array [
  Array [],
  Array [
    Object {
      "InstanceIds": Array [
        "IID",
      ],
    },
  ],
]
`;

exports[`aws autoscaling get can call get (instance ready) 1`] = `
Object {
  "instance": Object {
    "id": "IID",
    "ipaddress": "IPADDRESS",
    "launched": "LAUNCHTIME",
    "status": "STATUS",
  },
}
`;

exports[`aws autoscaling get can call get (instance ready) 2`] = `
Array [
  Array [],
  Array [
    Object {
      "AutoScalingGroupNames": Array [
        "ASG",
      ],
    },
  ],
  Array [],
  Array [
    Object {
      "AutoScalingGroupNames": Array [
        "ASG",
      ],
    },
  ],
  Array [],
  Array [
    Object {
      "AutoScalingGroupNames": Array [
        "ASG",
      ],
    },
  ],
]
`;

exports[`aws autoscaling get can call get (instance ready) 3`] = `
Array [
  Array [],
  Array [
    Object {
      "InstanceIds": Array [
        "IID",
      ],
    },
  ],
  Array [],
  Array [
    Object {
      "InstanceIds": Array [
        "IID",
      ],
    },
  ],
]
`;

exports[`aws autoscaling get can call get (no instances) 1`] = `
Object {
  "instance": null,
}
`;

exports[`aws autoscaling get can call get (no instances) 2`] = `
Array [
  Array [],
  Array [
    Object {
      "AutoScalingGroupNames": Array [
        "ASG",
      ],
    },
  ],
]
`;

exports[`aws autoscaling set should set desired capacity 1`] = `
Array [
  Array [
    Object {
      "AutoScalingGroupName": "ASG",
      "DesiredCapacity": 42,
    },
  ],
]
`;

exports[`aws lambda invoke should call invoke correctly 1`] = `
Array [
  Array [
    Object {
      "FunctionName": "ARN",
      "InvocationType": "Event",
      "Payload": "{\\"a\\":\\"A\\"}",
    },
  ],
]
`;

exports[`aws ssm getTelegramSecrets should cache 1`] = `
Object {
  "token": "TOKEN",
  "user": "USER",
}
`;

exports[`aws ssm getTelegramSecrets should call getParameters correctly 1`] = `
Object {
  "token": "TOKEN",
  "user": "USER",
}
`;

exports[`aws ssm getTelegramSecrets should call getParameters correctly 2`] = `
Array [
  Array [],
  Array [
    Object {
      "Names": Array [
        "NAME-telegram-token",
        "NAME-telegram-user",
      ],
      "WithDecryption": true,
    },
  ],
]
`;
